components/builder-depot/src/backend/s3.rs

```diff
@@ -310,3 +287,54 @@ pub struct S3Cli;
 impl Key for S3Cli {
     type Value = S3Handler;
 }
+
+// Helper function for programmatic creation of
+// the s3 object key
+fn s3_key(ident: &PackageIdent, target: &PackageTarget) -> Result<String> {
+    // Calling this method first ensures that the ident is fully qualified and the correct errors
+    // are returned in case of failure
+    let hart_name = ident.archive_name_with_target(target)?;
+
+    Ok(format!(
+        "{}/{}/{}",
+        ident.iter().collect::<Vec<&str>>().join("/"),
+        target.iter().collect::<Vec<&str>>().join("/"),
+        hart_name
+    ))
+}
```

Comment:
Is it my imagination, or is this returning a different kind of key from the function that it replaced?  It looks like the existing `s3_key` function returns a value of the form `X-X-X-X-X.hart` whereas this returns something that looks more like `X/X/X/X/X/X/X.hart`.  Is that going to break all of our artifact storage?  I don't understand why the format was changed.

Reply:
Good catch - it does indeed look like things will break if we don't keep the exact same format.

Reply:
It's a bit odd to look at, but the former version generated a key of the form:

```
$pkg_origin/$pkg_name/$pkg_version/$pkg_release/${pkg_arch}/${pkg_sys}/${pkg_origin}-${pkg_name}-${pkg_version}-${pkg_release}-${pkg_target}.hart
```

That is, there are 3 main pieces of information:

* The package ident parts, as a path, i.e. `${pkg_ident}`
* The package target parts, as a path, i.e. `${pkg_target}`
* The .hart file name, i.e. `${pkg_artifact}`

So, putting it another way:

```
${pkg_ident}/${pkg_target}/${pkg_artifact}
```

---

The updated implementation solves this with:

* `${pkg_ident}` paths are created using an iterator over `PackageIdent`
* `${pkg_target}` paths are created using an iterator over `PackageTarget`
* `${pkg_artifact}` entry is created calling the `archive_name_with_target` method on `PackageIdent`

You can see this in the unit test provided below, where I split the lead path entries from the filenaname and joined them in the `format!` call (the line was ludicrously long otherwise).

Make sense?

Reply:
To be clear, this should preserve the exact same key format

Reply:
👍